src/fileformats/ocd_georef_fields.cpp

```diff
@@ -0,0 +1,457 @@
+	        ? QString::number(result.crs_param) // common case, plain number
+	        : QString::number(-result.crs_param) + QLatin1String(" S"); // UTM South zones
+
+	for (const auto& spec_value : param->specValues(crs_param_string))
+		spec = spec.arg(spec_value);
+
+	// `spec` and the pair <`crs_id_string`, `crs_param_string`> both describe
+	// the same thing at this place
+	georef.setProjectedCRS(crs_id_string, spec, { crs_param_string });
+}
+
+/**
+ * OcdCrsRecord struct is used for passing CRS information around within the
+ * translation code.
+ */
+struct OcdCrsRecord
+{
+	const OcdGrid ocd_grid_id; ///< OCD grid number
+	const int ocd_zone_id; ///< OCD zone number, may be negative
+};
+
+/**
+ * Provide an OCD grid/zone combinations for a Mapper georeferencing.
+ *
+ * @see translateOcdToMapper
+ */
+OcdCrsRecord translateMapperToOcd(const MapperCrs crs_unique_id,
+                                  const int crs_param)
+{
+	switch (crs_unique_id)
+	{
+	case MapperCrs::Local:
+		return { OcdGrid::Local, 0 };
+		break;
+
+	case MapperCrs::Utm:
+		if (qAbs(crs_param) >= 1 && qAbs(crs_param) <= 60)
+			return { OcdGrid::Utm, crs_param };
+		break;
+
+	case MapperCrs::GaussKrueger:
+		if (crs_param >= 2 && crs_param <= 5)
+			return { OcdGrid::Germany, crs_param };
+		break;
+
+	case MapperCrs::Epsg:
+		// handle UTM separately, the rest is in the table
+		if (crs_param >= 32601 && crs_param <= 32660)
+			return { OcdGrid::Utm, crs_param - 32600 };
+		else if (crs_param >= 32701 && crs_param <= 32760)
+			return { OcdGrid::Utm, 32700 - crs_param };
+
+		for (const auto& record : grid_zone_epsg_codes)
+		{
+			if (record.epsg_code == crs_param)
+				return { record.ocd_grid_id, record.ocd_zone_id };
+		}
+		break;
+	default:
+		qWarning("Internal program error: translateMapperToOcd() should not"
+		         " be fed with invalid values (%d)", static_cast<int>(crs_unique_id));
+	}
+
+	return { OcdGrid::Invalid, 0 };
+}
+
+}  // anonymous namespace
+
+bool operator==(const OcdGeorefFields& lhs, const OcdGeorefFields& rhs)
+{
+	return lhs.i == rhs.i
+	        && lhs.m == rhs.m
+	        && lhs.x == rhs.x
+	        && lhs.y == rhs.y
+	        && ((qIsNaN(lhs.a) && qIsNaN(rhs.a)) || qAbs(lhs.a - rhs.a) < 5e-9) // 8-digit precision or both NaN's
+	        && lhs.r == rhs.r;
+}
+
+void OcdGeorefFields::setupGeoref(Georeferencing& georef,
+                                  const std::function<void (const QString&)>& warning_handler) const
+{
+	if (m >= 0)
+		georef.setScaleDenominator(m);
+
+	if (qIsFinite(a) && qAbs(a) >= 0.01)
+		georef.setGrivation(a);
+
+	if (r)
+		applyGridAndZone(georef, i, warning_handler);
+
+	QPointF proj_ref_point(x, y);
+	georef.setProjectedRefPoint(proj_ref_point, false);
+}
+
+OcdGeorefFields OcdGeorefFields::fromGeoref(const Georeferencing& georef,
+                                            const std::function<void (const QString&)>& warning_handler)
+{
+	OcdGeorefFields retval;
```

Comment:
a, m, x, y need to be set here: They can be returned and are useful even if everything else fails.

Reply:
They are already initialized via default member initialization to sensible defaults: https://github.com/lpechacek/mapper/blob/e12b60093dbb521c0e6063c1d50150026729016d/src/fileformats/ocd_georef_fields.h#L41-L46

Reply:
This is not about default values: It is about preserving as much input from OCD as possible when we hit an unknown grid. a, m, x, y will cause no trouble.

(Ideally, we would preserve the grid/zone parameter even if we don't understand the value. We might even record the adjustment to the CRS the user makes in Mapper so we can suggest this the next time we hit then unknown grid.)

Reply:
Parameter preservation is a new concept here. I'll have a look.